### DISCUSSION TITLE
Problems with Cron Jobs (Beta)

### DISCUSSION THREAD

**Discussion by lkenny97** [vercel/community#1919] [Help] [answered]
### Summary

Hello everyone,

I have 2 problems with Cron Jobs and wanted to discuss them here. Since the feature is in beta, I did not want to submit an issue or a bug report. So here is the discussion thread.

1. It is sad that there is no way to use environment variables in cron jobs. I have to write my key in my code and push it to my repo. Does anyone has a workaround for this problem? Something like this would fit nicely:
`{ "crons": [{ "path": "/api/test/${SECRET_KEY}", "schedule": "0 5 * * *" }] }`

2. The cron job does not work with custom domain for me and I can't seem to find any sources about this issue. I have a domain example.com (from godaddy) which I connected to my project in Vercel. I thought Vercel has 1 instance shared for all of my domains (depending on the branch). This does not seem to be the case since when I try to revalidate my pages using CRON, they do not get revalidated in example.com, but do in the unique domain Vercel generated for me. To reiterate, the problem is the following:

    1. I have my api here -> example.com/api/revalidatePages
    2. In my cron I set my path as /api/revalidatePages and expect example.com/pageThatNeedsRevalidation to be revalidated
    3. But the page that gets revalidated is some-unique-domain-vercel-generated.com/pageThatNeedsRevalidation
    4. If I send a GET request to example.com/api/revalidatePages everything works fine.

As far as I understand, the relative path sends the request to the wrong instance but there is no way to provide the domain because the documentation states that the path should be relative and start with a slash.

**Comment by amyegan** [comment, marked as answer]
There isn't a way to include secret keys yet, but the Cron team is working on it. 🙂 

The [deployment-specific URL](https://vercel.com/docs/concepts/deployments/generated-urls#) is used for cron jobs to ensure the invocation is bound to the specific deployment it was created with. If you're trying to revalidate based on a time schedule rather than at variable times when content changes, then [on-demand revalidation](https://nextjs.org/docs/basic-features/data-fetching/incremental-static-regeneration#on-demand-revalidation) might not be the best solution for you. Instead, you could set the timeout by adding the `revalidate` prop in `getStaticProps`. 

I was able to work around the domain difference by using a cron function to then call the revalidate function on my custom domain. But I'm not sure that's an ideal solution 😅

**Reply by lkenny97** [reply]
Looking forward to get the secret key feature 😄 

I guess I will have to steal your workaround. I do not want to use the `revalidate` prop because I want to update my page at 12 am to include the date in the DOM (without js for SEO). I could set a shorter period for revalidation but that page has other requests that are a bit heavy and I don't want to send them when it is not necessary.

**Comment by vvo** [comment]
Hey there, protecting Cron Jobs on Vercel is now possible using this method:

1. Create an environment variable named `CRON_SECRET` in your Vercel Project
2. When present, we will invoke your Cron Job with an authorization header this way: `Authorization: Bearer $CRON_SECRET`
3. You can use this authorization header to verify the Cron Job was invoked by Vercel

Let me know if that's working well. We will announce this more broadly soon, along with updating docs.
